xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/gl/Shader.h

```diff
@@ -0,0 +1,355 @@
+  virtual void Free() = 0;
+  virtual GLuint Handle() = 0;
+
+  bool LoadSource(const std::string& file)
+  {
+    char buffer[16384];
+
+    kodi::vfs::CFile source;
+    if (!source.OpenFile(file))
+    {
+      kodi::Log(ADDON_LOG_ERROR, "CShader::%s: Failed to open file '%s'", __FUNCTION__, file.c_str());
+      return false;
+    }
+    size_t len = source.Read(buffer, sizeof(buffer));
+    m_source.assign(buffer);
+    m_source[len] = 0;
+    source.Close();
+    return true;
+  }
+
+  bool OK() { return m_compiled; }
+
+protected:
+  std::string m_source;
+  std::string m_lastLog;
+  bool m_compiled = false;
+};
+//------------------------------------------------------------------------
+
+//========================================================================
+/// CVertexShader
+class ATTRIBUTE_HIDDEN CVertexShader : public CShader
+{
+public:
+  CVertexShader() = default;
+  ~CVertexShader() override { Free(); }
+
+  void Free() override
+  {
+    if (m_vertexShader)
+      glDeleteShader(m_vertexShader);
+    m_vertexShader = 0;
+  }
+
+  bool Compile(const std::string& extraBegin = "",
+                const std::string& extraEnd = "") override
+  {
+    GLint params[4];
+
+    Free();
+
+    m_vertexShader = glCreateShader(GL_VERTEX_SHADER);
+
+    GLsizei count = 0;
+    const char *sources[3];
+    if (!extraBegin.empty())
+      sources[count++] = extraBegin.c_str();
+    if (!m_source.empty())
+      sources[count++] = m_source.c_str();
+    if (!extraEnd.empty())
+      sources[count++] = extraEnd.c_str();
+
+    glShaderSource(m_vertexShader, count, sources, nullptr);
+    glCompileShader(m_vertexShader);
+    glGetShaderiv(m_vertexShader, GL_COMPILE_STATUS, params);
+    if (params[0] != GL_TRUE)
+    {
+      GLchar log[LOG_SIZE];
+      glGetShaderInfoLog(m_vertexShader, LOG_SIZE, nullptr, log);
+      kodi::Log(ADDON_LOG_ERROR, "CVertexShader::%s: %s", __FUNCTION__, log);
+      fprintf(stderr, "CVertexShader::%s: %s\n", __FUNCTION__, log);
+      m_lastLog = log;
+      m_compiled = false;
+    }
+    else
+    {
+      GLchar log[LOG_SIZE];
+      glGetShaderInfoLog(m_vertexShader, LOG_SIZE, nullptr, log);
+      m_lastLog = log;
+      m_compiled = true;
+    }
+    return m_compiled;
+  }
+
+  GLuint Handle() override { return m_vertexShader; }
+
+protected:
+  GLuint m_vertexShader = 0;
+};
+//------------------------------------------------------------------------
+
+//========================================================================
+/// CPixelShader
+class ATTRIBUTE_HIDDEN CPixelShader : public CShader
+{
+public:
+  CPixelShader() = default;
+  ~CPixelShader() { Free(); }
+  void Free() override
+  {
+    if (m_pixelShader)
+      glDeleteShader(m_pixelShader);
+    m_pixelShader = 0;
+  }
+
+  bool Compile(const std::string& extraBegin = "",
+                const std::string& extraEnd = "") override
+  {
+    GLint params[4];
+
+    Free();
+
+    if (m_source.length()==0)
```

Comment:
blank before and after '0'

Reply:
`if` becomes removed was a wrong leftover from old code.